src/cli.zig

```diff
@@ -1,78 +1,137 @@
-const std = @import("std");
+//! # Commands
+//! This library is built for parsing subcommand-based argument structures.
+//! Tha means there is an implicit parent command (ie the executable itself),
+//! which has zero or more subcommands, each of which itself may have zero or
+//! more subcommands themselves, recusively. If a (sub)command does not have
+//! any subcommands, it may accept positional arguments.
+//!
+//! For example, the command `foo --fizz="1" bar --buzz baz --fizzbuzz` can be broken
+//! down like so:
+//! ```zig
+//! .foo = .{
+//!     .fizz = "1",
+//!     .bar = .{
+//!         .buzz = true,
+//!         .baz = .{
+//!             .fizzbuzz = true,
+//!         },
+//!     },
+//! }
+//! ```
+//!
+//! # Options
+//! A given command may accept zero or more options, named arguments that may
+//! also be specified via an alias that is unique to it (per subcommand).
+//! An option may be of type bool, int, string, enum, and an optional xor
+//! list of any of the former primitives.
+//!
+//! Note: the `-h, --help` option cannot be used by the programmer, it is
+//! controlled by the parser.
+//!
+//! ## Optional Options, Required Options, Default Values
+//! All options must have a default value, but may be treated as "required" by
+//! the application by using an optional type with a default value of `null`,
+//! and issue an error if the option posseses this value.
+//!
+//! `null` is not a specifiable value for an optional on the command line itself,
+//! meaning if the value is `null`, the application can know for certain that the
+//! option was not specified at all.
+//!
+//! ## List Options
+//! List options are inherently "optional", in the sense that they are not allowed
+//! to have a non-empty default value, and will always default to an empty value
+//! if no values are specified. For this reason, a list cannot be optional, because
+//! whether or not the option was specified can be detected by whether or not the
+//! list is empty.
+//!
+//! Syntactically, lists are specified as a repeating series of the option.
+//! Example: `foo --bar 1 --bar 2 --bar 3 --bar 4`, would parse to a command
+//! where the option `bar` is equal to `&.{ 1, 2, 3, 4 }`.
+//!
+//! ## Positional Options
+//! A positional option is only allowed if there is no subcommand, and vice versa.
+//! They are specified before named options, and not after; for example:
+//! `foo file/path --opt=a`, but not `foo --opt=a file/path`.
 
-pub const CommandHelp = struct {
-    short: []const u8,
-    long: ?[]const u8,
-};
+const std = @import("std");
 
+// -- API -- //
+
+/// This is the core tool you will use to describe the shape of the command line parser.
```

Comment:
Isn't the struct definition the core tool that describes the shape? Isn't CommandInfo basically just for documentation and customization?

I just noticed that you can't even use the parser without an instance of CommandInfo, which seems odd. It would be nice if users of the library wouldn't be forced to deal with the extra complexity of creating a CommandInfo for every command, unless they want to add more details into the help message. It feels like a lot of redundant boilerplate.

Reply:
It's needed for disambiguating the interpretation of certain types, default value specification, as well as properly specifying subcommand info by extension.
It's quite a bit more complicated and hard to document without this. A previous design did in fact do such a thing, but it was unwieldy and felt very magical, in a bad way.

Reply:
In my opinion, the parser should have sane defaults that don't require you to include any `CommandInfo`s, but I think this is too complicated to figure out right now. It would probably involve some design changes, like specifying defaults on the structs themselves, which I know you saw some issues with. We can revisit this another time.